### PR TITLE
[memory][context] Global user access levels checks for complete actions in sc-memory and access levels checks to handle access levels for complete actions in sc-memory

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

Keynodes:

- nrel_user_action_class -- action class of user (класс действий пользователя);
- read_action_in_sc_memory -- action of reading in sc-memory (действие чтения в sc-памяти);
- write_action_in_sc_memory -- action of writing in sc-memory (действие записи в sc-памяти);
- erase_action_in_sc_memory -- action of erasing in sc-memory (действие удаления в sc-памяти);
- read_access_levels_action_in_sc_memory -- action of access levels reading in sc-memory (действие чтения прав доступа в sc-памяти);
- write_access_levels_action_in_sc_memory -- action of access levels writing in sc-memory (действие записи прав доступа в sc-памяти);
- erase_access_levels_action_in_sc_memory -- action of access levels erasing in sc-memory (действие удаления прав доступа в sc-памяти);

### CODE REVIEW

File: sc-memory/sc-core/sc_memory_context_manager.c [hunk 1]
```diff
@@ -156,29 +377,68 @@ void _sc_memory_context_manager_register_user_events(sc_memory_context_manager *
...
       manager,
       _sc_memory_context_manager_on_unauthentication_request_user,
       null_ptr);
+
+  sc_helper_resolve_system_identifier(
+      s_memory_default_ctx, "nrel_user_action_class", &manager->nrel_user_action_class_addr);
+  _sc_context_set_access_levels_for_element(
+      manager->nrel_user_action_class_addr, SC_CONTEXT_ACCESS_LEVEL_TO_ALL_ACCESS_LEVELS);
+
+  sc_context_manager_add_basic_action_class_access_levels("read_action_in_sc_memory", SC_CONTEXT_ACCESS_LEVEL_READ);
```
Comment (KovalM) [inline sc-memory/sc-core/sc_memory_context_manager.c:408]: I think it's time for us to store the names of all the keynodes on which sc-memory depends in one file

File: sc-memory/sc-core/sc_memory_context_manager.c [hunk 2]
```diff
@@ -37,102 +46,310 @@ struct _sc_event_emit_params
...
+
+/**
+ * @brief Adds basic action class access levels to the sc-memory context manager.
+ * @param _action_class_system_idtf System identifier of the action class.
+ * @param _access_levels Access levels to be added for the action class.
+ */
+#define sc_context_manager_add_basic_action_class_access_levels(_action_class_system_idtf, _access_levels) \
+  ({ \
+    sc_addr _action_class_addr; \
+    sc_helper_resolve_system_identifier(s_memory_default_ctx, _action_class_system_idtf, &_action_class_addr); \
```
Comment (KovalM) [inline sc-memory/sc-core/sc_memory_context_manager.c:240]: Default access classes will be created without class type

File: sc-memory/sc-core/sc_memory_context_manager.c [hunk 3]
```diff
@@ -37,102 +46,310 @@ struct _sc_event_emit_params
...
+
+  return SC_RESULT_OK;
+}
+
+/**
+ * @brief Adds basic action class access levels to the sc-memory context manager.
+ * @param _action_class_system_idtf System identifier of the action class.
+ * @param _access_levels Access levels to be added for the action class.
+ */
+#define sc_context_manager_add_basic_action_class_access_levels(_action_class_system_idtf, _access_levels) \
```
Comment (KovalM) [inline sc-memory/sc-core/sc_memory_context_manager.c:237]: Why we can add action class by system idtf but get it we can by sc-addr? It turns out to be inconsistency

File: sc-memory/sc-core/sc_memory_context_manager.c [hunk 4]
```diff
@@ -37,102 +46,310 @@ struct _sc_event_emit_params
...
 {
-  sc_unused(&arc_addr);
+  sc_unused(&initiator_addr);
+
+  // Only temporal sc-arcs can be used
+  if (sc_type_check(connector_type, sc_type_arc_pos_const_temp) == SC_FALSE)
+    return SC_RESULT_OK;
+
+  _sc_context_set_access_levels_for_element(connector_addr, SC_CONTEXT_ACCESS_LEVEL_TO_ALL_ACCESS_LEVELS);
+  _sc_context_set_access_levels_for_element(user_addr, SC_CONTEXT_ACCESS_LEVEL_TO_ERASE_ACCESS_LEVELS);
```
Comment (KovalM) [inline sc-memory/sc-core/sc_memory_context_manager.c:169]: I couldn’t understand this, it needs to be discussed

**Review by KovalM — CHANGES_REQUESTED** [review]
1. During watching the pull request, I didn’t understand well why we need to add a check for reading, writing or erasing at all now, if in the future we want to make access rights differently: each user will be able to read, write or delete, just in different structures. Now I don’t see how the checks from this pull request will lead us to our goal.
2. I have an idea about the need for transactions, we also need to discuss this

**Comment by NikitaZotov** [comment]
> 1. During watching the pull request, I didn’t understand well why we need to add a check for reading, writing or erasing at all now, if in the future we want to make access rights differently: each user will be able to read, write or delete, just in different structures. Now I don’t see how the checks from this pull request will lead us to our goal.
> 2. I have an idea about the need for transactions, we also need to discuss this

It is necessary when you want to say that user can read everything in the knowledge base, but cannot edit something in some local structure

File: sc-memory/sc-core/sc_memory_context_manager.c (hunk 2 above)
Comment (NikitaZotov) [reply sc-memory/sc-core/sc_memory_context_manager.c:240]: I think sc-machine doesn't respond for this. Types can be different. As for me, I think that 
1. simple set types are more appropriate than class types for nodes from which temporal arcs are outgoing
2. semantic types can removed from sc-machine in the future

File: sc-memory/sc-core/sc_memory_context_manager.c (hunk 3 above)
Comment (NikitaZotov) [reply sc-memory/sc-core/sc_memory_context_manager.c:237]: It is private define that hides the same logic to create keynodes

File: sc-memory/sc-core/sc_memory_context_manager.c (hunk 4 above)
Comment (NikitaZotov) [reply sc-memory/sc-core/sc_memory_context_manager.c:169]: There are settings of access levels for keynodes that used to set access levels for users. It is needed to check after that we can add arcs from it

File: sc-memory/sc-core/sc_memory_context_manager.c (hunk 1 above)
Comment (NikitaZotov) [reply sc-memory/sc-core/sc_memory_context_manager.c:408]: Why?

File: sc-memory/tests/sc-memory/common/test_sc_memory_context.cpp [hunk 5]
```diff
@@ -0,0 +1,902 @@
+#include <gtest/gtest.h>
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/tests/sc-memory/common/test_sc_memory_context.cpp:1]: This file is changed fully. So it is no need to change its name in another PR

[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/tests/sc-memory/common/test_sc_memory_context.cpp:1]: 250 lines -> 900 lines

File: sc-memory/sc-memory/sc_memory.cpp [hunk 6]
```diff
@@ -206,6 +206,11 @@ size_t ScMemoryContext::GetElementOutputArcsCount(ScAddr const & addr) const
     SC_THROW_EXCEPTION(
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-memory/sc_memory.cpp:206]: ok

File: sc-tools/sc-server/tests/api/units/test_users.cpp [hunk 7]
```diff
@@ -13,6 +13,56 @@
...
+  ScAddr const & nrelUserActionClassAddr = context->HelperFindBySystemIdtf("nrel_user_action_class");
+  ScAddr const & edgeAddr = context->CreateEdge(ScType::EdgeDCommonConst, userAddr, actionClassAddr);
+  context->CreateEdge(ScType::EdgeAccessConstPosTemp, nrelUserActionClassAddr, edgeAddr);
+}
+
+void TestAddAccessLevelsForUserToInitReadActions(
+    std::unique_ptr<ScMemoryContext> const & context,
+    ScAddr const & userAddr)
+{
+  ScAddr const & readActionInScMemoryAddr = context->HelperFindBySystemIdtf("read_action_in_sc_memory");
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-tools/sc-server/tests/api/units/test_users.cpp:30]: You can write propositions here

[the remark this answers is not in the document]
Comment (KovalM) [reply sc-tools/sc-server/tests/api/units/test_users.cpp:30]: action_read_from_sc_memory, action_generate_in_sc_memory, action_erase_from_sc_memory

**Review by KovalM — CHANGES_REQUESTED** [review]
two comments are still actual

**Review by KovalM — APPROVED** [review]

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: sc-memory/sc-core/sc_memory.c
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc_memory.c:1]: Done

File: sc-memory/sc-core/sc-store/sc_iterator3.c
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_iterator3.c:1]: Done

File: sc-memory/sc-core/sc-store/sc_element.h [hunk 8]
```diff
@@ -47,7 +47,7 @@ struct _sc_arc_info
 struct _sc_element_flags
 {
   sc_type type;
-  sc_access_levels access_levels;
+  sc_states states;
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_element.h:50]: I rename it after Kovalev's review. We think that it is appropriate name

[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-core/sc-store/sc_element.h:50]: This field stores not only access levels, but lists of freed elements, life flags and etc.

**Review by MksmOrlov — APPROVED** [review]

**Pull request merged**